STYLE.md

```diff
@@ -0,0 +1,453 @@
+is doing, you should add a comment.
+* Always provide a type signature
+* Use Haddock syntax in the comments.
+* Comment every exported data type.
+
+Function example:
+
+```haskell
+-- | Send a message on a socket. The socket must be in a connected
+-- state. Returns the number of bytes sent. Applications are
+-- responsible for ensuring that all data has been sent.
+send
+  :: Socket      -- ^ Connected socket
+  -> ByteString  -- ^ Data to send
+  -> IO Int      -- ^ Bytes sent
+```
+
+* For functions, the documentation should give enough information to
+apply the function without looking at its definition.
+
+Record example:
+
+```haskell
+-- | Bla bla bla.
+data Person = Person
+  { age :: !Int     -- ^ Age
+  , name :: !String  -- ^ First name
+  }
+```
+
+For fields that require longer comments, format them this way:
+
+```haskell
+data Record = Record
+  { -- | This is a very very very long comment that is split over
+    -- multiple lines.
+    field1 :: !Text
+
+    -- | This is a second very very very long comment that is split
+    -- over multiple lines.
+  , field2 :: !Int
+  }
+```
+
+#### End-of-Line Comments
+
+Separate end-of-line comments from the code with 2 spaces. Align
+comments for data type definitions. Some examples:
+
+```haskell
+data Parser =
+  Parser
+    !Int         -- ^ Current position
+    !ByteString  -- ^ Remaining input
+
+foo :: Int -> Int
+foo n = salt * 32 + 9
+  where
+    salt = 453645243  -- Magic hash salt.
+```
+
+#### Links
+
+Use in-line links economically.  You are encouraged to add links for
+API names. It is not necessary to add links for all API names in a
+Haddock comment. We therefore recommend adding a link to an API name
+if:
+
+* The user might actually want to click on it for more information (in
+  your opinion), and
+
+* Only for the first occurrence of each API name in the comment (don't
+  bother repeating a link)
+
+Top-down guideline
+------------------
+
+### LANGUAGE extensions section
+
+Write each `LANGUAGE` pragma on its own line, sort them alphabetically.
+_Do not_ align by max width as future re-alignment when adding new pragmas
+will cause semantically null changes in the diff.
+
+```haskell
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE TemplateHaskell #-}
+```
+
+### Module name
+
+Use singular when naming modules (e.g. use `Data.Map` and
+`Data.ByteString.Internal` instead of `Data.Maps` and
+`Data.ByteString.Internals`). Sometimes it's acceptable to use plural
+(e. g. `Types`, `Instances`).
+
+### Export Lists
+
+Format export lists as follows:
+
+```haskell
+module Data.Set
+  ( Set
+  , empty
+  , singleton
+  , member
+  ) where
+```
+
+Some clarifications:
+
+1. Use 2 spaces indentation for export list.
+2. You can split export list into sections or just write all as single section.
+3. It is strongly adviced to sort each section alpabetically. However,
+   classes, data types and type aliases should be written before
+   functions.
```

Comment:
I think there's another angle to the story:

1. The export list could reflect the order of definitions, so as to give reader an expectation of where they shall find what they seek for.
2. The definition order often has its own logic.

Reply:
I personally like the alphabetical ordering, grouped as types followed by functions and then optionally functions exported purely for testing (with a comment).

I do however recognize that as purely my own preference that I would have difficulty justifying over the alternatives.

Reply:
> I personally like the alphabetical ordering, grouped as types followed by functions and then optionally functions exported purely for testing (with a comment).

I also like this, @deepfire  what do you think?

Reply:
There is yet another angle -- sometimes module exports several themes of entities, that:
1. Related enough to come together in a single module,
2. Yet distinct enough to merit their own grouping in the export list.

Like, for example in: https://github.com/input-output-hk/cardano-node/blob/b783a886d3afc76326165dbaeb54e671d3feab48/cardano-node/src/Cardano/CLI/Key.hs#L7

Sometimes those subgroups are crystallising towards a separation into a module of their own.
In this case having the group clearly visible helps towards that goal.

It's absolutely worth keeping each group sorted alphabetically, with types coming first, though!

What do you think, @erikd, @Jimbo4350, @intricate  ?

Reply:
I've never felt like I'd gained much from the grouping of exports. I can see the value of dividing the exports if some are used only in testing. If I want to know what a function does (or group of functions), I'll look at said function & comment to go along with it. This is strictly a preference however.

I would personally opt for:
- Alphabetical ordering, grouped as types followed by functions and then optionally functions exported purely for testing (with a comment).

Reply:
@deepfire: I don't mind the grouping of exports. I rarely do so myself but I've seen a few places where it was nice (no examples, unfortunately.. can't remember right now).

Reply:
>Sometimes those subgroups are crystallising towards a separation into a module of their own.
In this case having the group clearly visible helps towards that goal.

This seems like a fair point. I wouldn't be opposed to this type of grouping, then. However, I probably wouldn't strictly enforce it.